tests/test_client_errors.py

```diff
@@ -52,16 +65,45 @@ def close(self):
 
 
 class MockTransport(v3io.dataplane.transport.httpclient.Transport):
-    def __init__(self, *args, connection_options=None, **kwargs):
+    def __init__(self, *args, connection_options=None, reset_after_create_connections=False, **kwargs):
         self.mock_connections = []
         self.connection_options = connection_options or {}
+        self.reset_after_create_connections = reset_after_create_connections
         super().__init__(*args, **kwargs)
 
     def _create_connection(self, host, ssl_context):
         conn = MockConnection(**self.connection_options)
         self.mock_connections.append(conn)
         return conn
 
+    def _create_connections(self, num_connections, host, ssl_context):
+        super()._create_connections(num_connections=num_connections, host=host, ssl_context=ssl_context)
+        if self.reset_after_create_connections:
+            self.connection_options = {}
+
+
+def test_first_connection_failure():
+    connection_options = {"fail_request_after": 0, "exception_class": CannotSendRequest}
+    client = v3io.dataplane.Client()
+
+    mock_transport = MockTransport(
+        client._logger, connection_options=connection_options, reset_after_create_connections=True
+    )
+    client._transport = mock_transport
+    size = 1024
+    data = os.urandom(size)
+    with mmap.mmap(-1, size) as mmap_obj, tempfile.NamedTemporaryFile(mode="w+b", delete=False) as temp_file:
+        mmap_obj.write(data)
+        mmap_obj.seek(0)
+        components = temp_file.name.split("/")
+        container = components[1]  # Index 0 will be an empty string due to leading '/'
+        path = "/" + "/".join(components[2:])
+        client.put_object(
+            container=container, path=path, body=mmap_obj, raise_for_status=v3io.dataplane.RaiseForStatus.never
+        )
+        temp_file.seek(0)
+        assert temp_file.read() == data
```

Comment:
Normally the error speaks for itself, but in this case it really doesn't.
```suggestion
        assert temp_file.read() == data, "Binary data read back differs from original data"
```